xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp

```diff
@@ -1365,6 +1365,7 @@ void CActiveAE::Configure(AEAudioFormat *desiredFmt)
     // buffers for viz
     if (!(inputFormat.m_dataFormat == AE_FMT_RAW))
     {
+      CSingleLock lock(m_vizLock);
```

Comment:
the lock is there to protect audiocallback. makes no sense to grab this lock over some hundred lines of code.

Reply:
I disagree with that assessment.  The lock could be tighter, but it's only needlessly locking a few lines of code certainly not hundreds of lines of code, the lock will be released when it falls out of scope on line 1392.  

m_audioCallback's state is checked as part of the if() statement on line 1376.  It seems excessive to add a scope just to lock and grab a boolean and then test it when locking m_audioCallback for a few extra milliseconds isn't going to hurt anything.

I am happy to remove the added line of code if you prefer, it was just something that appeared to be an oversight in the implementation.

I would also like your thoughts on if the entire change is too much of a hack since all it's doing is masking an issue somewhere else by preventing a condition that likely shouldn't need to be prevented at all.

Reply:
You are right, I didn't look at the diff thoroughly. That is even a different function than the other changes.

> condition that likely shouldn't need to be prevented at all.

The root cause is here: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L2454
The error>0 block checks for number of samples being zero https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L2445 but the error<0 block does not.

Reply:
Awesome, thank you! Let me try that and see if it has the desired effect.  I thought there were a couple other scenarios I ran into where it could get zero samples all the way to the visualizer, but you're almost undoubtedly correct (as expected).

If modifying the error < 0 block does the trick, do you think it prudent to also leave the test I added to ensure a bum packet can't be sent to an attached visualization? It would probably be better to remove it and point out the concern with the individual visualization module(s)?

Thanks again, I feel like I waste a lot of your time.

Reply:
Yeah, that seems to do the trick. Let me come up with something moderately clean and I'll update.  First iteration is frankly just too ugly to submit, calling pop_front() from within SyncStream(), which causes far too many side-effects to adjust for.  Maybe tomorrow :)